src/graph/hungarian-algorithm.md

```diff
@@ -0,0 +1,291 @@
+
+**Step 2.** Further, at each step of the algorithm, we try, without changing the potential, to increase the cardinality of the current matching $M$ by one (recall that the matching is searched in the graph of rigid edges $H$). To do this, the usual [Kuhn's Algorithm for finding the maximum matching in bipartite graphs](kuhn_maximum_bipartite_matching.md) is used. Let us recall the algorithm here.
+All edges of the matching $M$ are oriented in the direction from the right part to the left one, and all other edges of the graph $H$ are oriented in the opposite direction.
+
+Recall (from the terminology of searching for matchings) that a vertex is called saturated if an edge of the current matching is adjacent to it. A vertex that is not adjacent to any edge of the current matching is called unsaturated. A path of odd length, in which the first edge does not belong to the matching, and for all subsequent edges there is an alternating belonging to the matching (belongs/does not belong) - is called an augmenting path.
+From all unsaturated vertices in the left part, a [depth-first](depth-first-search.md) or [breadth-first](breadth-first-search.md) traversal is started. If, as a result of the search, it was possible to reach the top unsaturated vertex of the right part, we have found an augmenting path from the left part to the right one. If we include odd edges of the path and remove the even ones in the matching (i.e. include the first edge in the matching, exclude the second, include the third, etc.), then we will increase the matching cardinality by one.
+
+If there was no augmenting path, then the current matching $M$ is maximal in the graph $H$.
+
+**Step 3.** If at the current step, it is not possible to increase the cardinality of the current matching, then a recalculation of the potential is performed in such a way that, at the next steps, there will be more opportunities to increase the matching.
+
+Denote by $Z_1$ the set of vertices of the left part that were visited during the last traversal of Kuhn's algorithm, and through $Z_2$ the set of visited vertices of the right part.
+
+Let's calculate the value $\Delta$:
+
+$$\Delta = \min_{i\in Z_1,\ j\notin Z_2} A[i][j]-u[i]-v[j].$$
+
+!!! info ""
+
+     **Lemma.** $\Delta > 0.$
+
+??? info "Proof"
+
+    Suppose $\Delta=0$. Then there exists a rigid edge $(i,j)$ with $i\in Z_1$ and $j\notin Z_2$. It follows that the edge $(i,j)$ must be oriented from the right part to the left one, i.e. $(i,j)$ must be included in the matching $M$. However, this is impossible, because we could not get to the saturated vertex i except by going along the edge from j to i. So $\Delta > 0$.
+
+Now let's **recalculate the potential** in this way:
+
+- for all vertices $i\in Z_1$, do $u[i] \gets u[i]+\Delta$,
+
+- for all vertices $j\in Z_2$, do $v[j] \gets v[j]-\Delta$.
+
+Note that conditions are not mutually exclusive.
```

Comment:
I'm sorry, which conditions not mutually exclusive with what?

Reply:
I think this sentence was from a deleted paragraph or similar. I deleted it.